Wikipedia/Code/SavedArticlesFetcherProgressManager.swift

```diff
@@ -0,0 +1,63 @@
+/*
+ This class acts as a bridge between the SavedArticlesFetcher (which retrieves article html and images) and VC's which want to display save article fetching progress. It does so while being minimally invasive to SavedArticlesFetcher, observing only the SavedArticlesFetcher's 'fetchesInProcessCount'. Interested VC's can access the SavedArticlesFetcherProgressManager.shared.progress - but be aware that this "progress" object is not re-used, so you'll need to observe it to know when it gets re-set, and you'll then have to observe any properties you are interested in of the re-set progress. Reason:
+ 
+ "NSProgress objects cannot be reused. Once they’re done, they’re done. Once they’re cancelled, they’re cancelled. If you need to reuse an NSProgress, instead make a new instance and provide a mechanism so the client of your progress knows that the object has been replaced, like a notification." ( Source: https://developer.apple.com/videos/play/wwdc2015/232/ by way of https://stinkykitten.com/index.php/2017/08/13/nsprogress/ )
+*/
+
+@objcMembers class SavedArticlesFetcherProgressManager: NSObject, ProgressReporting {
+
+    static let shared = SavedArticlesFetcherProgressManager()
```

Comment:
Is there only one saved articles fetcher? If so, couldn't this just be a property on that class? it seems tightly coupled to that class so it would make sense. Also would get rid of the this being shared and make it easier to kill.

Reply:
Yes I struggled with this.

The SavedArticlesFetcher is currently only init'ed by the app view controller, but there's more than one thing that needs to have a way to know about the saved article fetching progress:
- the progress view (currently using it) 
- the icon badge (currently only reflecting the reading list API activity - this was one of the follow-up issues - Josh is aware and we'll prob chat about it at meeting tomorrow)

So making the one instance of the SavedArticlesFetcher accessible to these other areas means we'd need a way to expose it. I was reluctant to make the whole thing just a singleton for this purpose, and my initial attempt to relay it otherwise were (way) more invasive, so I settled on this manager which was solely responsible for exposing (via `shared` and conformance to `ProgressReporting`) and managing as much progress logic as it could.

I can take a stab again at exposing the SavedArticlesFetcher instead - perhaps I should just make it a singleton?

Reply:
Got it… so you are saying that progress needs to be known by several objects, but those objects don't necessarily need to be able to control the fetcher - is that right?

In that case what sticks out to me is that the fetcher is just global anyways… it is a variable on the progress manager. And the progress manager is global. 

So by that logic… if the appVC nils out it's fetcher, the progress manager would still be retaining the old one. So that feels strange to me. 

Basically in the current setup, the progress manager is kinda controlling the life cycle of the fetcher… which feels inverted to me.

Honestly, the progress feels like it could be owned by the fetcher. I mean for the most part I don't think anything else in the system even needs to know about the progress manager class except for the fetcher. If the fetcher just exposed the resulting progress of the manager as a var that could be monitored, it could hide the complexity of the manager class from the rest of the world - which may be nice. 

But given your other comments… lets just say that the progress manager is now a variable of the fetcher and any object that has access to the fetcher can monitor the progress… do you actually need to make the fetcher global? Can it be passed by dependency injection instead? Or is it just to "far away" to do cleanly?

Reply:
> Got it… so you are saying that progress needs to be known by several objects, but those objects don't necessarily need to be able to control the fetcher - is that right?

Yes I believe so.

> So by that logic… if the appVC nils out it's fetcher, the progress manager would still be retaining the old one.n Basically in the current setup, the progress manager is kinda controlling the life cycle of the fetcher… which feels inverted to me.

Ah, good point on the inversion/lifecycle, though presently the appVC doesn't appear to ever nil out the fetcher. 

> If the fetcher just exposed the resulting progress of the manager as a var that could be monitored, it could hide the complexity of the manager class from the rest of the world - which may be nice.

That does sound better, though I'll have to check again on the "too far away" issue you mentioned. I'll take a peek...

Reply:
Looking at the view hierarchy this may not be so bad... taking a stab at it...